In [2]:
import numpy as np
import random

In [33]:
inpImage = np.random.rand(227,227,3)
receptiveWindow = np.random.rand(11,11,3)
inpImage

array([[[ 0.69356189,  0.69440331,  0.93070711],
        [ 0.77977364,  0.68585178,  0.35657396],
        [ 0.87892536,  0.94328858,  0.01802211],
        ..., 
        [ 0.91653314,  0.04898516,  0.3719315 ],
        [ 0.08741859,  0.92646525,  0.45894163],
        [ 0.72646241,  0.51932963,  0.5973866 ]],

       [[ 0.92025818,  0.71938475,  0.06747324],
        [ 0.71131298,  0.84807078,  0.8799578 ],
        [ 0.54233422,  0.90405989,  0.03459637],
        ..., 
        [ 0.26639783,  0.06636401,  0.688515  ],
        [ 0.21426611,  0.83356741,  0.86449363],
        [ 0.88859151,  0.29800399,  0.64573658]],

       [[ 0.75202288,  0.71215628,  0.92358507],
        [ 0.04549526,  0.91412978,  0.26253996],
        [ 0.34273941,  0.98553372,  0.94988396],
        ..., 
        [ 0.7648293 ,  0.13919293,  0.28531005],
        [ 0.95121185,  0.06448176,  0.10939861],
        [ 0.00617377,  0.45805022,  0.20671783]],

       ..., 
       [[ 0.79113812,  0.73434291,  0.75696037],
        

In [34]:
receptiveWindow

array([[[ 0.23247431,  0.72904587,  0.04767989],
        [ 0.71047751,  0.60606048,  0.15801135],
        [ 0.08705429,  0.35437736,  0.23983709],
        [ 0.8168822 ,  0.58678233,  0.90940338],
        [ 0.29472847,  0.99395335,  0.62460917],
        [ 0.1869425 ,  0.34715329,  0.98489901],
        [ 0.63073282,  0.26962992,  0.01702188],
        [ 0.55484526,  0.95300577,  0.78436909],
        [ 0.93221147,  0.39084261,  0.20934666],
        [ 0.76815594,  0.39122003,  0.4212079 ],
        [ 0.41491232,  0.17860726,  0.85360298]],

       [[ 0.0790626 ,  0.30772248,  0.2406031 ],
        [ 0.51068656,  0.9775372 ,  0.11628941],
        [ 0.31992911,  0.23037038,  0.61132436],
        [ 0.00607921,  0.57233836,  0.7534785 ],
        [ 0.44072709,  0.08154959,  0.09122829],
        [ 0.83540113,  0.04017301,  0.49262054],
        [ 0.79784171,  0.43800785,  0.53984645],
        [ 0.48466566,  0.44011274,  0.79423286],
        [ 0.1009591 ,  0.63536279,  0.96948368],
        [ 0.992199

In [37]:
receptiveWindow = receptiveWindow.flatten()

In [38]:
print (len(receptiveWindow))

363


In [44]:
# Defining Hyperparameters:
d = 96     # d is the depth of the output volume, It can be interpreted as the number of the filters used for the model
w = 227    # w is the width of the input image
f = 11     # f is the width of the filter or receptive window
s = 4      # s is the number of strid 
p = 0      # p is the number of zero-padding


# Shape of the output activatation layer is given as: []
x_dir = (w-f + (2*p))/(s) + 1
y_dir = (w-f + (2*p))/(s) + 1
activationShape = [x_dir,y_dir]
print (activationShape)

[55.0, 55.0]


Normally we slide the receptive window of size (11x11x3) over the Image of size [227x227x3] using a stride of 4. The receptive field flattened would have a size of 11*11*3= 363. The receptive field has to slide through the image with a stride of 4, which would give a total activationShape of [55x55] flattened to 55*55=3025 cells. Finally we have the image to column form as [363x3025]. Here the rows are the receptive window flattened and columns are the flattened activation window activated by a receptive field. imCols = [363x3025]

On the other hand we have weights that is the total shape of the filter plus 1 bias term. For this case not not assume the bias term. For a case where we share weights we would have a weight matrix which is again in the shape of the receptive field, which is [11x11x3]=363, For each filter we will have a weigth matrix of (96x363) different type of filter of width of filter we will have. wRow = [96x363]

The final output activationShape would be of np.dot(wRow, imCols) = [96x3025]
The output activationShape find the activated neuron output for each filter. This is also called as the Actication Map.